src/SignalR/common/Http.Connections/src/Internal/HttpConnectionContext.cs

```diff
@@ -170,6 +176,13 @@ public IDuplexPipe Application
 
         public HttpContext HttpContext { get; set; }
 
+        public override CancellationToken ConnectionClosed { get; set; }
+
+        public override void Abort()
+        {
+            ThreadPool.UnsafeQueueUserWorkItem(cts => ((CancellationTokenSource)cts).Cancel(), _connectionClosedTokenSource);
```

Comment:
Abort should be using HttpContext.Abort (if it's not the fake HttpContext). So even if you keep this, it should result in killing the transport directly.

Reply:
I'm not sure if calling HttpContesxt.Abort() is the way to do it, but I agree this should also stop the transport if it hasn't stopped already.

Reply:
> this should also stop the transport if it hasn't stopped already.

That's the part I don't have currently

Reply:
Is there a race with HttpContext.Abort and Kestrel writing the bytes in the response stream to the network? As soon as I added this it started randomly failing tests that were looking for a close message.

Reply:
I wouldn't call it a race as much as behaving as intended. Aborting drops any unflushed bytes in Kestrel's internal buffers. If you want to guarantee delivery of the bytes you've written, you cannot abort. This is the same reason why we cannot reliably abort a SignalR connection and send a CloseMessage.

> I'm not sure if calling HttpContesxt.Abort() is the way to do it

Apparently it isn't.

Reply:
So do we need to abort the HttpContext? Or can we assume the server will clean up correctly when we return from the middleware?

Reply:
The server should do the right thing to complete the request and/or close the connection when middleware completes. If the client receives data slower the configured rate limit, the server will abort the connection.